pkg/controller/hibernation/openstack_actuator.go

```diff
@@ -0,0 +1,1091 @@
+		logger.Infof("all instances are being deleted (%v) - hibernation in progress", deletingInstances)
+		return false, []string{"instances-being-deleted"}, nil
+	}
+
+	return runningCount > 0, []string{}, nil
+}
+
+// Check the actual state of instances in OpenStack
+func (a *openstackActuator) categorizeInstanceStates(openstackClient openstackclient.Client, servers []*servers.Server, logger log.FieldLogger) (int, []string) {
+	runningCount := 0
+	var deletingInstances []string
+
+	for _, server := range servers {
+		status := strings.ToLower(server.Status)
+		if strings.Contains(status, "delet") || status == "shutoff" || status == "error" {
+			logger.Infof("instance %s is being deleted/stopped (status: %s)", server.Name, server.Status)
+			deletingInstances = append(deletingInstances, server.Name)
+		} else {
+			logger.Infof("instance %s is running (status: %s)", server.Name, server.Status)
+			runningCount++
+		}
+	}
+
+	return runningCount, deletingInstances
+}
+
+// Check if machines are stopped
+func (a *openstackActuator) MachinesStopped(cd *hivev1.ClusterDeployment, hiveClient client.Client, logger log.FieldLogger) (bool, []string, error) {
+	logger = logger.WithField("cloud", "openstack")
+	logger.Info("checking if machines are stopped")
+
+	openstackClient, err := a.openstackClientFn(cd, hiveClient, logger)
+	if err != nil {
+		return false, nil, fmt.Errorf("failed to create OpenStack client: %v", err)
+	}
+
+	infraID := cd.Spec.ClusterMetadata.InfraID
+	matchingServers, err := a.findInstancesByInfraID(openstackClient, infraID)
+	if err != nil {
+		return false, nil, fmt.Errorf("error finding instances: %v", err)
+	}
+
+	if len(matchingServers) == 0 {
+		logger.Info("no instances found - machines are stopped")
+		return true, nil, nil
+	}
+
+	var notStopped []string
+	for _, server := range matchingServers {
+		notStopped = append(notStopped, server.Name)
+	}
+
+	logger.Infof("found %d instances still running", len(notStopped))
+	return false, notStopped, nil
+}
+
+type OpenStackInstanceConfig struct {
+	Name               string            `json:"name"`
+	Flavor             string            `json:"flavor"`
+	PortID             string            `json:"portID"`
+	SnapshotID         string            `json:"snapshotID"`
+	SnapshotName       string            `json:"snapshotName"`
+	SecurityGroups     []string          `json:"securityGroups"`
+	ClusterID          string            `json:"clusterID"`
+	NetworkID          string            `json:"networkID"`
+	OpenshiftClusterID string            `json:"openshiftClusterID"`
+	Metadata           map[string]string `json:"metadata"`
+	Tags               []string          `json:"tags"`
+}
+
+// Return servers that match the infraID prefix
+func (a *openstackActuator) findInstancesByInfraID(openstackClient openstackclient.Client, prefix string) ([]*servers.Server, error) {
+	ctx := context.Background()
+
+	allServers, err := openstackClient.ListServers(ctx, nil)
```

Comment:
I think I mentioned this [previously](https://github.com/openshift/hive/pull/2722#discussion_r2274304095): can we prefilter using `Tags`?

Reply:
I'm suggesting that we can use the `ListOpts` param to ListServers() to filter the results on the server side, thus reducing our resource usage (network traffic, client-side filtering). Again, I'm not positive how (or even if!) servers are tagged by installer, but they are for other platforms. So my suggestion would be to do something like:
```suggestion
	allServers, err := openstackClient.ListServers(ctx, &servers.ListOpts{
		// I would rename `prefix` to `infraID`
		Tags: prefix,
	})
```
...and then you could get rid of LL615-20.